### PR TITLE
Adding Basic Support for Windows Docker Images

### PR DESCRIPTION
**Pull request by nowens03**
**container_import**
Added a parameter called “manifest” and made the “layers” parameter optional.  This allows Bazel to import a Windows base image without checking in GBs of tar files.  All that is needed is the config and manifest files from the base Windows image.

container_import(
    name = "windowsservercore_1803",
    config = "windowsservercore.1803.config.json",
    manifest = "windowsservercore.1803.manifest.json"
)

**container_image**
Added an optional parameter to docker_build called “operating_system.”  While most of the changes are foreign layer specific, two changes were required to address Windows specifically.

There is no support for adding registry diffs, but diffs in existing layers will be respected and continue to work.

container_image(
    name = "basic_windows_image",
    base = ":import_windows_base_image",
    cmd = ['bar.exe'],
    files = [":bar.exe"]
    operating_system = "windows",
)

### CODE REVIEW

File: container/image.bzl [hunk 1]
```diff
@@ -168,6 +179,13 @@ def _image_config(
         args += ["--base=%s" % base_config.path]
         inputs += [base_config]
 
+    if base_manifest:
+      args += ["--basemanifest=%s" % base_manifest.path]
+      inputs += [base_manifest]
+
+    if operating_system:
+      args += ["--operating_system=%s" % operating_system]
+
```
Comment (nlopezgi) [inline container/image.bzl]: it seems you're not using the --architecture flag here

File: container/import.bzl [hunk 2]
```diff
@@ -83,6 +83,13 @@ def _container_import_impl(ctx):
         blobsums += [_sha256(ctx, zipped)]
         diff_ids += [_sha256(ctx, unzipped)]
 
+    manifest = ""
+    manifest_digest = ""
+
+    if not (len(ctx.files.manifest) == 0):
```
Comment (nlopezgi) [inline container/import.bzl]: if (len(ctx.files.manifest) > 0): (to make consistent with check in 136)

File: container/import.bzl [hunk 3]
```diff
@@ -83,6 +83,13 @@ def _container_import_impl(ctx):
         blobsums += [_sha256(ctx, zipped)]
         diff_ids += [_sha256(ctx, unzipped)]
 
+    manifest = ""
```
Comment (nlopezgi) [inline container/import.bzl]: should these be set to None?

File: container/import.bzl [hunk 4]
```diff
@@ -140,7 +156,8 @@ def _container_import_impl(ctx):
 container_import = rule(
     attrs = dict({
         "config": attr.label(allow_files = [".json"]),
-        "layers": attr.label_list(allow_files = tar_filetype + tgz_filetype),
+        "manifest": attr.label(allow_files = [".json"], mandatory = False),
+        "layers": attr.label_list(allow_files = tar_filetype + tgz_filetype, mandatory = False),
```
Comment (nlopezgi) [inline container/import.bzl]: why is layers now optional. I saw the comment on the PR that states "This allows Bazel to import a Windows base image without checking in GBs of tar files" but I'm not sure I understand why having layers for windows image import is problematic (why it results in checking in GBs of tar files). Also, would it not be valid to just pass an empty list for layers?

File: container/layer.bzl [hunk 5]
```diff
@@ -84,6 +85,13 @@ def build_layer(
         "--mode=" + ctx.attr.mode,
     ]
 
+    if (operating_system == 'windows'):
```
Comment (nlopezgi) [inline container/layer.bzl]: Could you document what (and why) you are doing here

File: container/layer.bzl [hunk 6]
```diff
@@ -194,6 +206,8 @@ _layer_attrs = dict({
     # Implicit/Undocumented dependencies.
     "empty_files": attr.string_list(),
     "empty_dirs": attr.string_list(),
+    "architecture": attr.string(default="amd64", mandatory = False),
```
Comment (nlopezgi) [inline container/layer.bzl]: does not seem like you are using this architecture attr anywhere in the impl?

File: tests/docker/configs/windows_image.yaml [hunk 7]
```diff
@@ -0,0 +1,4 @@
+schemaVersion: 2.0.0
+
+metadataTest:
+  cmd: ["echo bar"]
```
Comment (nlopezgi) [inline tests/docker/configs/windows_image.yaml]: Could you add structure tests validating the empty root directories were successfully created?

File: container/build_tar.py [hunk 8]
```diff
@@ -39,6 +39,11 @@
 gflags.DEFINE_string(
     'mode', None, 'Force the mode on the added files (in octal).')
 
+gflags.DEFINE_multistring(
```
[the remark this answers is not in the document]
Comment (nowens03) [reply container/build_tar.py]: I'm updating the doc as follows:

    'An empty root directory to add to the layer.  This will create a directory that'
    'is a peer of "root_directory".  "empty_dir" creates an empty directory inside of'
    '"root_directory"'

In Linux layers, the root directory is named '.'  In windows there are two root directories 'Files' (the equivalent of '.') and 'Hives'.  Hives is where the registry deltas would are stored.  The directory needs to be present for the image to load, but there doesn't need to be anything populated.

File: container/build_tar.py [hunk 9]
```diff
@@ -188,6 +201,23 @@ def add_empty_dir(self, destpath, mode=None, ids=None, names=None):
     self.add_empty_file(destpath, mode=mode, ids=ids, names=names,
                         kind=tarfile.DIRTYPE)
 
+  def add_empty_root_dir(self, destpath, mode=None, ids=None, names=None):
+    """Add a directory to the root of the tar file.
```
[the remark this answers is not in the document]
Comment (nowens03) [reply container/build_tar.py]: This is for the 'Hives' directory.  It needs to be present at the root of the image, but does not need to have any content.  In the future if someone added support for registry deltas in the Windows layer they would be put here.

File: container/build_tar.py [hunk 10]
```diff
@@ -77,6 +82,9 @@
                           'Specify the owner names of individual files, e.g. '
                           'path/to/file=root.root.')
 
+gflags.DEFINE_string(
+    'root_directory', './', 'Default root directory is named "."')
```
[the remark this answers is not in the document]
Comment (nowens03) [reply container/build_tar.py]: Windows Docker image layers store the files in a root directory called "Files" instead of one named "." like Linux.

Updated documentation to reflect this also.

File: container/build_tar.py [hunk 11]
```diff
@@ -375,6 +405,8 @@ def file_attributes(filename):
       output.add_empty_file(f, **file_attributes(f))
     for f in FLAGS.empty_dir:
       output.add_empty_dir(f, **file_attributes(f))
+    for f in FLAGS.empty_root_dir:
```
[the remark this answers is not in the document]
Comment (nowens03) [reply container/build_tar.py]: In Windows there are two root directories ('Files', 'Hives') instead of one as with Linux images ('.').  These changes are actually brought over from the bazel repo where build_tar.py was forked from at some point.

In the context of the bazel repo build_tar.py isn't necessarily Docker specific so following the existing pattern of empty_dir and allowing multiple empty root directories makes sense.

In this repo I could modify this so it only adds the empty root directory needed ('Hives' since 'Files' will store all the data that Linux images normally store in '.').  If you'd like me to constrain this to 'Hives' I can.  Please let me know what approach you'd prefer.

File: container/create_image_config.py [hunk 12]
```diff
@@ -76,10 +82,11 @@
...
                     help='If True, "Cmd" will be set to null.')
 
-_PROCESSOR_ARCHITECTURE = 'amd64'
-
-_OPERATING_SYSTEM = 'linux'
+parser.add_argument('--architecture', action='store', default='amd64',
+                    help=('Architecture to create docker image for, e.g. {amd64}'))
 
+parser.add_argument('--operating_system', action='store', default='linux',
+                    help=('Operating system to create docker image for, e.g. {linux}'))
```
[the remark this answers is not in the document]
Comment (nowens03) [reply container/create_image_config.py]: Done

File: container/create_image_config.py [hunk 13]
```diff
@@ -31,9 +31,15 @@
 parser.add_argument('--base', action='store',
                     help='The parent image.')
 
+parser.add_argument('--basemanifest', action='store',
+                    help='The parent image.')
```
[the remark this answers is not in the document]
Comment (nowens03) [reply container/create_image_config.py]: Fixed

File: container/build_tar.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (nlopezgi) [reply container/build_tar.py]: I like it how it is right now. Do not know enough about windows filesystem and your comments are very helpful. Your other changes to docs so far look good, with the additional context its clear this should allow multiple empty_root_dir (also, like this is generic as opposed to hardcoded for windows).

File: container/build_tar.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (nlopezgi) [reply container/build_tar.py]: got it, was just suggesting comment at the top could be 
"Add an empty directory to the root of the tar file"

File: container/build_tar.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (nlopezgi) [reply container/build_tar.py]: awesome, this clears things up, will look at the updated docs.

File: container/build_tar.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (nlopezgi) [reply container/build_tar.py]: thanks for clarifying. lets try to get this info about 'hives' somewhere in the code docs (like in container/layer.bzl ~ line 88)

File: container/join_layers.py [hunk 14]
```diff
@@ -199,22 +256,34 @@ def main():
...
+      unzipped_filename, zipped_filename) = elts
+
+      diff_id = 'sha256:' + utils.ExtractValue(diffid_filename)
+      blob_sum = 'sha256:' + utils.ExtractValue(blobsum_filename)
+
+      diffid_to_blobsum[diff_id] = blob_sum
+      blobsum_to_unzipped[blob_sum] = unzipped_filename
+      blobsum_to_zipped[blob_sum] = zipped_filename
+
+  # add foreign layers
```
[the remark this answers is not in the document]
Comment (nowens03) [reply container/join_layers.py]: I've updated the source with the following:

  # add foreign layers
  #
  # Windows base images distributed by Microsoft are using foreign layers.
  # Foreign layers are not stored in the Docker repository like normal layers.
  # Instead they include a list of URLs where the layer can be downloaded.
  # This is done because Windows base images are large (2+GB).  When someone
  # pulls a Windows image, it downloads the foreign layers from those URLs
  # instead of requesting the blob from the registry.
  # When adding foreign layers through bazel, the actual layer blob is not
  # present on the system.  Instead the base image manifest is used to
  # describe the parent image layers.

**Comment by nlopezgi** [comment]
So one alternative I was thinking for the forked container/archive.py file, is that we could add an http_file (example: https://github.com/bazelbuild/bazel-toolchains/blob/master/repositories/repositories.bzl#L177) to pull https://github.com/bazelbuild/bazel/blob/55dc53f441847bbfc9abe2ef65696fc722e7cfec/tools/build_defs/pkg/archive.py and then have a BUILD file under third_party/BUILD with the py_library rule to build the archive.py file. Would still mean we need to change this once Bazel release is out, but at least we do not need to check in the fork. (the downside of this approcach is we will likely have to do some copybara changes when sync'ing rules_docker to internal repos) wdyt?

File: container/join_layers.py [hunk 15]
```diff
@@ -174,18 +241,8 @@ def main():
...
-    (fq_tag, config_filename) = elts
-
-    formatted_tag = fq_tag.format(**stamp_info)
-    tag = docker_name.Tag(formatted_tag, strict=False)
-    config_file = utils.ExtractValue(config_filename)
-
-    # Add the mapping in one direction.
-    tag_to_config[tag] = config_file
+  tag_to_config = create_tag_to_file(stamp_info, args.tags)
+  tag_to_manifest = create_tag_to_file(stamp_info, args.manifests)
```
[the remark this answers is not in the document]
Comment (nowens03) [reply container/join_layers.py]: Agreed, this is confusing.  I changed the function signature to look like this:

def create_tag_to_file_content_map(stamp_info, tag_file_pairs):
  """
    Creates a Docker image tag to file content map.

    Args:
      stamp_info - Tag substitutions to make in the input tags, e.g. {BUILD_USER}
      tag_file_pairs - List of input tags and file names
          (e.g. ...:image=@bazel-out/...image.0.config)
  """

I think it's more clear now.  Please let me know if you'd like more done here.

File: .bazelci/presubmit.yml [hunk 16]
```diff
@@ -35,6 +35,26 @@ platforms:
...
+    - "..."
+    - "-//docker/..."
+    - "-//skylib/..."
+    - "-//tools/..."
+    test_flags:
+    - "--copt=-w"
+    - "--host_copt=-w"
+    test_targets:
+    # We are running the skipped targets remotely only.
+    - "--"
```
Comment (nlopezgi) [inline .bazelci/presubmit.yml]: Thanks for adding this config.
Lets just build / test the specific windows tests you just created on windows

**Comment by nlopezgi** [comment]
As I thought, windows presubmit failed as likely most of those targets cannot run on windows. Lets just run the new ones we want to verify in that config.

**Comment by nlopezgi** [comment]
travis CI failed due to buildifier issues, please run buildifier on the files you changed.

File: container/create_image_config.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (nowens03) [reply container/create_image_config.py]: Done

File: container/image.bzl [hunk 17]
```diff
@@ -556,6 +591,11 @@ def container_image(**kwargs):
   Args:
     **kwargs: See above.
   """
+    operating_system = None
+
+    if ("operating_system" in kwargs):
+        operating_system = kwargs["operating_system"]
```
[the remark this answers is not in the document]
Comment (nowens03) [reply container/image.bzl]: Done

File: container/image.bzl [hunk 18]
```diff
@@ -440,9 +472,12 @@ container_image_ = rule(
 #       "a"
 #   ],
 # NOTE: prefacing a command with 'exec' just ends up with the former
-def _validate_command(name, argument):
+def _validate_command(name, argument, operating_system):
```
[the remark this answers is not in the document]
Comment (nowens03) [reply container/image.bzl]: Done

File: container/create_image_config.py [hunk 19]
```diff
@@ -76,10 +82,11 @@
 parser.add_argument('--null_cmd', action='store', default=False,
                     help='If True, "Cmd" will be set to null.')
 
-_PROCESSOR_ARCHITECTURE = 'amd64'
-
-_OPERATING_SYSTEM = 'linux'
+parser.add_argument('--architecture', action='store', default='amd64',
```
[the remark this answers is not in the document]
Comment (nowens03) [reply container/create_image_config.py]: It's not actually used right now.  Docker does support other architectures though so it may make sense going forward.  Would you like me to revert this part of the change?

File: container/image.bzl (hunk 1 above)
Comment (nowens03) [reply container/image.bzl]: I'm not, but it may be useful to plumb it all the way through for support for arm, etc.  What would you prefer?

File: container/import.bzl (hunk 2 above)
Comment (nowens03) [reply container/import.bzl]: Done

File: container/layer.bzl (hunk 6 above)
Comment (nowens03) [reply container/layer.bzl]: Plumbed it all the way through now to create_image_config.py.

File: container/image.bzl (hunk 1 above)
Comment (nowens03) [reply container/image.bzl]: Looks like v2.MetadataTest doesn't understand architecture yet in the image config file.  I've manually verified that it's working, but it's not as good as a test.

I'll need to update the test project as well to make this work.  Would you like me to remove architecture altogether or bring another PR later with the test working?

File: .bazelci/presubmit.yml [hunk 20]
```diff
@@ -35,6 +35,19 @@ platforms:
     - "--"
     - "..."
     - "-//tests/docker/..."
+  windows:
+    build_flags:
+    - "--copt=-w"
+    - "--host_copt=-w"
+    build_targets:
+    # We are running the skipped targets remotely only.
```
Comment (nlopezgi) [inline .bazelci/presubmit.yml]: no need to have this comment here or below

File: container/BUILD [hunk 21]
```diff
@@ -64,14 +64,21 @@ py_binary(
     ],
 )
 
+py_library(
```
Comment (nlopezgi) [inline container/BUILD]: could you please try the alternative of having a http_file to pull the archive.py file from Bazel repo?

File: container/join_layers.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (nlopezgi) [reply container/join_layers.py]: lgtm

**Comment by nlopezgi** [comment]
@nowens03 
wrt "I'll need to update the test project as well to make this work. Would you like me to remove architecture altogether or bring another PR later with the test working?"

I think its best to remove architecture altogether from this PR and introduce it later when the test is working.

File: container/image.bzl (hunk 18 above)
[the remark this answers is not in the document]
Comment (nlopezgi) [reply container/image.bzl]: looks like you have not yet pushed the change for this docs

File: container/image.bzl (hunk 17 above)
[the remark this answers is not in the document]
Comment (nlopezgi) [reply container/image.bzl]: lgtm

**Comment by nowens03** [comment]
I've been building these all on a Linux machine since that is our use case.  I'm not super familiar with Bazel, but I believe some of the external are not yet available for Windows (e.g. Rust).  I've removed the presubmit for Windows for now.  I'll take another look at what it will take to get rules_docker building on Windows.

**Comment by nlopezgi** [comment]
Change looks good. @xingao267 can you provide advice as to how to fix the cla/google check (@nowens03 , I think you might need to recreate the PR to get that check passing).
Once all checks are passing we can approve and merge.
Thanks for all the changes @nowens03 and thanks a lot for sending this PR, we love that rules_docker can now support windows images!

**Comment by nowens03** [comment]
Re-creating to pass cla/google check. I didn't have my google.com email account associated with my github account.  I've updated that.

**Comment by xingao267** [comment]
@nowens03  You also need to have your git config user.email in your machine set to your google.com email.
